File: analysis/src/data/get_census_data.R
pacman::p_load(tidyverse, 
               tidycensus, 
               sf, 
               optparse,
               tidyjson, 
               jsonlite, 
               purrr)


option_list = list(
    make_option(c("--config"), type="character", default=NULL, 
              help="configuration file", metavar="character"),
    make_option(c("--city"), type="character", default=NULL, 
              help="city code", metavar="character")    
); 
 
opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

config <- fromJSON(txt=opt$config)
city <- opt$city
base_path <- config$base_path
config <- map_dfr(config[city], as_tibble)
msa_code <- config[config$city_code==city,]$msa_code

path = paste0(base_path,"/cities/",city,"/census/geo/")
tract_path = "tracts.geojson"
tract_2010_path = "tracts_2010.geojson"
block_group_path = "block_groups.geojson"
# block_path = "geo/blocks.geojson"
# acs_path =  "acs/acs_wide.csv"
print(path)
dir.create(path)

all_msa <- read_csv("https://www.bls.gov/cew/classifications/areas/qcew-county-msa-csa-crosswalk-csv.csv") # National MSA to county crosswalk
selected_msa <- all_msa %>% filter(`MSA Code` == msa_code)
selected_msa_counties <- selected_msa %>% pull(`County Code`)
selected_msa_states <- selected_msa %>% separate(`County Title`, sep = ",", into = c("name", "state")) %>%  pull(state) %>% unique()

## Save 2020 Tract boundaries
tract_boundaries <- get_acs(geography = "tract",
                            state = selected_msa_states,
                            year = 2020, 
                            cb = TRUE, 
                            geometry = T, 
                            variables = "B19013_001") %>% 
  mutate(county = substr(GEOID, 1,5)) %>%
  filter(county %in% selected_msa_counties) %>%
  select(-variable, -estimate, -moe)

tract_boundaries <- st_transform(tract_boundaries, "WGS84")


st_write(tract_boundaries, paste0(path, tract_path),append=FALSE)

## Save 2010 Tract boundaries
tract_boundaries <- get_acs(geography = "tract",
                            state = selected_msa_states,
                            year = 2010, 
                            cb = TRUE, 
                            geometry = T, 
                            variables = "B19013_001") %>% 
  mutate(county = substr(GEOID, 1,5)) %>%
  filter(county %in% selected_msa_counties) %>%
  select(-variable, -estimate, -moe)

tract_boundaries <- st_transform(tract_boundaries, "WGS84")


st_write(tract_boundaries, paste0(path, tract_2010_path),append=FALSE)

## Save 2020 Block groups
block_group_boundaries <- get_acs(geography = "block group", 
                                  state = selected_msa_states, 
                                  year = 2019, 
                                  cb = TRUE, 
                                  geometry = T,
                                  variables = "B19013_001") %>% 
  mutate(county = substr(GEOID, 1,5)) %>%
  filter(county %in% selected_msa_counties) %>%
  select(-variable, -estimate, -moe)

block_group_boundaries <- st_transform(block_group_boundaries, "WGS84")

st_write(block_group_boundaries, paste0(path, block_group_path),append=FALSE)

File: analysis/src/data/get_census_data.py
from optparse import OptionParser
import pandas as pd
import geopandas as gpd
import json
from pathlib import Path
import os
import censusdis.data as ced
from pandas import json_normalize


def get_census(config, city_key):
    """
    Retreives census shapefiles for city based on msa codes supplied in config
    """

    base_path = config['base_path']
    city_config = json_normalize(config[city_key])
    msa_codes = city_config[city_config['city_code'] == city_key]['msa_code'].iloc[0]

    # File Paths
    path = f"{base_path}/cities/{city_key}/census/geo/"
    msa_path = f"{base_path}/national/qcew-county-msa-csa-crosswalk-csv.csv"
    tract_path = "tracts.geojson"
    tract_2010_path = "tracts_2010.geojson"
    block_group_path = "block_groups.geojson"    

    if not os.path.isdir(path):
        os.makedirs(path)
    
    all_msa = pd.read_csv(msa_path, encoding="ISO-8859-1")
    selected_msa = all_msa[all_msa['MSA Code'].isin(msa_codes)]
    selected_msa_counties_five_digits = selected_msa['County Code'].astype(str).str.zfill(5)
    selected_msa_states = list(set(state[:2] for state in selected_msa_counties_five_digits))    

    # 2020 block groups
    gdf_block_group = ced.download("acs/acs5", 2020, ["NAME"], state=selected_msa_states, county = "*", block_group = "*", with_geometry=True)

    gdf_block_group["GEOID"] = gdf_block_group["STATE"] + gdf_block_group["COUNTY"] + gdf_block_group["TRACT"] + gdf_block_group["BLOCK_GROUP"]
    gdf_block_group["county"] = gdf_block_group["STATE"] + gdf_block_group["COUNTY"]
    gdf_block_group = gdf_block_group[["GEOID", "NAME", "county", "geometry"]]
    gdf_block_group = gdf_block_group.query("county.isin(@selected_msa_counties_five_digits)")

    gdf_block_group.to_file(f'{path}{block_group_path}')

    # 2020 tracts
    gdf_tract = ced.download("acs/acs5", 2020, ["NAME"], state=selected_msa_states, county = "*", tract = "*", with_geometry=True)

    gdf_tract["GEOID"] = gdf_tract["STATE"] + gdf_tract["COUNTY"] + gdf_tract["TRACT"] 
    gdf_tract["county"] = gdf_tract["STATE"] + gdf_tract["COUNTY"]
    gdf_tract = gdf_tract[["GEOID", "NAME", "county", "geometry"]]
    gdf_tract = gdf_tract.query("county.isin(@selected_msa_counties_five_digits)")

    gdf_tract.to_file(f'{path}{tract_path}')

    # 2010 tracts
    gdf_tract_2010 = ced.download("acs/acs5", 2010, ["NAME"], state=selected_msa_states, county = "*", tract = "*", with_geometry=True)

    gdf_tract_2010["GEOID"] = gdf_tract_2010["STATE"] + gdf_tract_2010["COUNTY"] + gdf_tract_2010["TRACT"] 
    gdf_tract_2010["county"] = gdf_tract_2010["STATE"] + gdf_tract_2010["COUNTY"]
    gdf_tract_2010 = gdf_tract_2010[["GEOID", "NAME", "county", "geometry"]]
    gdf_tract_2010 = gdf_tract_2010.query("county.isin(@selected_msa_counties_five_digits)")

    gdf_tract_2010.to_file(f'{path}{tract_2010_path}')

def main():
    parser = argparse.ArgumentParser("Get Census")
    parser.add_argument("--config", required=True)
    parser.add_argument("--city", required=True)
    
    opts = parser.parse_args()
    
    with open(opts.config) as f:
        config = json.load(f)

    get_census(config, opts.city)
    print("Census geographies written") 
    
if __name__ == "__main__":
    main()